fcitx-contrib/fcitx5-macos
# Notifications

<img width="354" alt="image" src="https://github.com/fcitx-contrib/fcitx5-macos/assets/23358293/1f11feb1-aff7-435f-9b1d-e8ba2a44a2eb">


<img width="355" alt="image" src="https://github.com/fcitx-contrib/fcitx5-macos/assets/23358293/077aef96-7409-47ef-9e1f-fa994d427cb2">


This PR introduces a new addon `macosnotifications` that overrides the standard `notifications` addon to support notifications using the native `UserNotifications` framework. This addon is statically linked into the main program, similar to `macosfrontend`. (Unfortunately, the main program must be modified to request authorization to send notifications in `applicationDidFinishLaunching`.)

There are three APIs to support:

- `sendNotification`
- `showTip`
- `closeNotification`

This PR is a complete implementation, modulo `appIcon` (impossible afaiu). "Actions" and "timeouts" are supported. For example, if you click on "Do not show again", this type of notifications won't show again.

Note:

- Must enable codesigning or notifications will be denied by macOS. Configure with `cmake -DENABLE_CODESIGN=ON -DCODESIGN_KEY=<your codesign key>`. A self-signed key should suffice for testing purposes.

## How to test

1. Make sure `Fcitx5` is codesigned, and grant the notification permission. `install-name-tool` will invalidate the codesign, so don't use the installed binary.
2. Remove `pinyin.conf`, restart fcitx5, and then get a notification asking whether to enable cloudpinyin.
3. With fcitx5-rime, there are tips for deployment and toggling simplification, half-shape, etc. (After clicking "Do not show again", the hidden tipIds will be saved in `~/.config/fcitx5/conf/macosnotifications.conf`.)
